### PR TITLE
Ensure python tooling requires hashes, bump builders trusty->xenial

### PR DESCRIPTION
**Pull request by msheiny**
This change helps maintain that each developer is pulling the exact same
resources, help mitigate against nefarious actors in the pip pipeline,
and solve male baldness.

## Status

Work in Progress

## Description of Changes

Kinda, sorta addresses #3270 .
Doesn't do anything to squash the multiple hashes for some packages tho.

Changes proposed in this pull request:

* Breaks out pip update make target into a script
* Adds hashes and removes the banner on those generated requirements files.
* Replaces trusty with xenial for the builder containers. Why did I do this? Well I encountered a fun chicken before the egg problem with pip. Specifically I wanted to utilize hashes but the pip version installed in trusty does not support hashes (its really old, like 1.x series). It would have been super silly to peg all hashes and then just wildly run `pip install -U pip` before that. Sooooo I made the call to jump the builders to xenial which has a much newer version of pip pre-installed.
* Fixes some misc ansible errors

On a side note, our quay Clare scanner reported dramatically less vulnerabilities when using xenial vs. trusty. Take that report with a grain of salt but hey i consider this a win.

## Testing

How should the reviewer test this PR?

* Make sure CI passes :+1: 
* Make sure you can build debs and that `vagrant up /staging/` comes up.

## Deployment

Any special considerations for deployment?

This doesn't dramatically change deployment story but it does update a lot of the tooling behind the builder.

### CODE REVIEW

File: devops/scripts/update-pip-reqs.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+#
+# Needs to be run from repo root directory
```
Comment (heartsucker) [inline devops/scripts/update-pip-reqs.sh]: You can add this to ensure it's always run at the repo root. Portable across Linux/Mac.

```bash
get_script_dir() {
  declare src="${BASH_SOURCE[0]}"
  declare dir=

  while [ -h "$src" ]; do
    dir="$(cd -P "$( dirname "$src")" && pwd)"
    src="$(readlink "$src")"
    [[ $src != /* ]] && src="$dir/$src"
  done
  cd -P "$(dirname "$src")" && pwd 
}
cd "$(get_script_dir)/../../"
```

File: install_files/ansible-base/roles/build-ossec-deb-pkg/tasks/main.yml [hunk 2]
```diff
@@ -52,9 +52,10 @@
         SHA1: {{ ossec_download.stat.checksum }}
 
       Try rerunning the playbook to download the files again.
-  when: not (ossec_download.stat.exists and
-             ossec_download.stat.md5 == "{{ ossec_md5_checksum }}" and
-             ossec_download.stat.checksum == "{{ ossec_sha1_checksum }}")
+  when:
+    - not ossec_download.stat.exists
```
Comment (heartsucker) [inline install_files/ansible-base/roles/build-ossec-deb-pkg/tasks/main.yml]: This logic isn't the same. You need to add a `!=` for the other two to get the correct negations.

File: molecule/builder/Dockerfile [hunk 3]
```diff
@@ -1,9 +1,9 @@
-# April 28th, 2018 ubuntu:trusty
-FROM ubuntu@sha256:b8855dc848e2622653ab557d1ce2f4c34218a9380cceaa51ced85c5f3c8eb201
+# ubuntu:xenial-20180417
```
Comment (heartsucker) [inline molecule/builder/Dockerfile]: 👍

**Comment by heartsucker** [comment]
> Doesn't do anything to squash the multiple hashes for some packages tho.

This is intentional since there's like eggs and wheels and all sorts of other ways to actually install the software. Look at the bottom of the page for `cryptography` on [PyPI](https://pypi.org/project/cryptography/#files).

**Comment by heartsucker** [comment]
On a fresh checkout

```
vagrant destroy -f /staging/
rm -rf venv
virtualenv -p `which python2` venv
source venv/bin/activate
pip install -r securedrop/requirements/develop-requirements.txt
make build debs
vagrant up /staging
```

gives me this on `.deb` install:

```
failed: [app-staging] (item=securedrop-app-code-0.8.0~rc1-amd64.deb) => {"changed": true, "cmd": ["dpkg", "-i", "/root/securedrop-app-code-0.8.0~rc1-amd64.deb"], "delta": "0:00:01.176646", "end": "2018-05-25 09:08:38.402801", "item": "securedrop-app-code-0.8.0~rc1-amd64.deb", "msg": "non-zero return code", "rc": 1, "start": "2018-05-25 09:08:37.226155", "stderr": "debconf: unable to initialize frontend: Dialog\ndebconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.)\ndebconf: falling back to frontend: Readline\ndebconf: unable to initialize frontend: Readline\ndebconf: (This frontend requires a controlling tty.)\ndebconf: falling back to frontend: Teletype\n+ pip install --no-index --find-links=/var/securedrop/wheelhouse --upgrade -r /var/www/securedrop/requirements/securedrop-app-code-requirements.txt\ndpkg: error processing package securedrop-app-code (--install):\n subprocess installed post-installation script returned error exit status 2\nErrors were encountered while processing:\n securedrop-app-code", "stderr_lines": ["debconf: unable to initialize frontend: Dialog", "debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.)", "debconf: falling back to frontend: Readline", "debconf: unable to initialize frontend: Readline", "debconf: (This frontend requires a controlling tty.)", "debconf: falling back to frontend: Teletype", "+ pip install --no-index --find-links=/var/securedrop/wheelhouse --upgrade -r /var/www/securedrop/requirements/securedrop-app-code-requirements.txt", "dpkg: error processing package securedrop-app-code (--install):", " subprocess installed post-installation script returned error exit status 2", "Errors were encountered while processing:", " securedrop-app-code"], "stdout": "(Reading database ... 44568 files and directories currently installed.)\nPreparing to unpack .../securedrop-app-code-0.8.0~rc1-amd64.deb ...\nUnpacking securedrop-app-code (0.8.0~rc1) over (0.8.0~rc1) ...\nSetting up securedrop-app-code (0.8.0~rc1) ...\nIgnoring indexes: https://pypi.python.org/simple/\nException:\nTraceback (most recent call last):\n  File \"/usr/lib/python2.7/dist-packages/pip/basecommand.py\", line 122, in main\n    status = self.run(options, args)\n  File \"/usr/lib/python2.7/dist-packages/pip/commands/install.py\", line 262, in run\n    for req in parse_requirements(filename, finder=finder, options=options, session=session):\n  File \"/usr/lib/python2.7/dist-packages/pip/req.py\", line 1632, in parse_requirements\n    req = InstallRequirement.from_line(line, comes_from, prereleases=getattr(options, \"pre\", None))\n  File \"/usr/lib/python2.7/dist-packages/pip/req.py\", line 173, in from_line\n    return cls(req, comes_from, url=url, prereleases=prereleases)\n  File \"/usr/lib/python2.7/dist-packages/pip/req.py\", line 71, in __init__\n    req = pkg_resources.Requirement.parse(req)\n  File \"/usr/lib/python2.7/dist-packages/pkg_resources.py\", line 2667, in parse\n    reqs = list(parse_requirements(s))\n  File \"/usr/lib/python2.7/dist-packages/pkg_resources.py\", line 2605, in parse_requirements\n    line, p, specs = scan_list(VERSION,LINE_END,line,p,(1,2),\"version spec\")\n  File \"/usr/lib/python2.7/dist-packages/pkg_resources.py\", line 2583, in scan_list\n    \"Expected ',' or end-of-list in\",line,\"at\",line[p:]\nValueError: (\"Expected ',' or end-of-list in\", 'asn1crypto==0.24.0 \\\\', 'at', ' \\\\')\n\nStoring debug log for failure in /root/.pip/pip.log", "stdout_lines": ["(Reading database ... 44568 files and directories currently installed.)", "Preparing to unpack .../securedrop-app-code-0.8.0~rc1-amd64.deb ...", "Unpacking securedrop-app-code (0.8.0~rc1) over (0.8.0~rc1) ...", "Setting up securedrop-app-code (0.8.0~rc1) ...", "Ignoring indexes: https://pypi.python.org/simple/", "Exception:", "Traceback (most recent call last):", "  File \"/usr/lib/python2.7/dist-packages/pip/basecommand.py\", line 122, in main", "    status = self.run(options, args)", "  File \"/usr/lib/python2.7/dist-packages/pip/commands/install.py\", line 262, in run", "    for req in parse_requirements(filename, finder=finder, options=options, session=session):", "  File \"/usr/lib/python2.7/dist-packages/pip/req.py\", line 1632, in parse_requirements", "    req = InstallRequirement.from_line(line, comes_from, prereleases=getattr(options, \"pre\", None))", "  File \"/usr/lib/python2.7/dist-packages/pip/req.py\", line 173, in from_line", "    return cls(req, comes_from, url=url, prereleases=prereleases)", "  File \"/usr/lib/python2.7/dist-packages/pip/req.py\", line 71, in __init__", "    req = pkg_resources.Requirement.parse(req)", "  File \"/usr/lib/python2.7/dist-packages/pkg_resources.py\", line 2667, in parse", "    reqs = list(parse_requirements(s))", "  File \"/usr/lib/python2.7/dist-packages/pkg_resources.py\", line 2605, in parse_requirements", "    line, p, specs = scan_list(VERSION,LINE_END,line,p,(1,2),\"version spec\")", "  File \"/usr/lib/python2.7/dist-packages/pkg_resources.py\", line 2583, in scan_list", "    \"Expected ',' or end-of-list in\",line,\"at\",line[p:]", "ValueError: (\"Expected ',' or end-of-list in\", 'asn1crypto==0.24.0 \\\\', 'at', ' \\\\')", "", "Storing debug log for failure in /root/.pip/pip.log"]}
```

**Comment by msheiny** [comment]
yeahh thanks for the early review @heartsucker ... its obviously not ready for merge right now ... i also encountered breaking tests during local development that i wasnt able to fix. if someone else wants to step in and help out here :heart: :heart:  otherwise im going to jump to some other tasks for a few days. I kind of did this ticket on a whim and I thought it would be easy-peasy...that older pip issue ended up causing a lot of headache and will require more work.

**Comment by redshiftzero** [comment]
now that we're on xenial only for the 0.13.x release series, closing in favor of #4435 :heart:

**Pull request closed without merging**